src/GTP.cpp

```diff
@@ -516,18 +516,30 @@ bool GTP::execute(GameState & game, std::string xinput) {
     } else if (command.find("heatmap") == 0) {
         std::istringstream cmdstream(command);
         std::string tmp;
-        int rotation;
+        std::string rotation;
+        Network::Netresult vec;
 
         cmdstream >> tmp;   // eat heatmap
         cmdstream >> rotation;
 
-        if (!cmdstream.fail()) {
-            auto vec = Network::get_scored_moves(
-                &game, Network::Ensemble::DIRECT, rotation, true);
-            Network::show_heatmap(&game, vec, false);
+        if (!cmdstream.fail() && rotation == "all") {
+            for (int r=0; r<8; r++) {
+                vec = Network::get_scored_moves(
+                    &game, Network::Ensemble::DIRECT, r, true);
+                Network::show_heatmap(&game, vec, false);
+            }
         } else {
-            auto vec = Network::get_scored_moves(
-                &game, Network::Ensemble::DIRECT, 0, true);
+            if (cmdstream.fail()) {
+                vec = Network::get_scored_moves(
+                    &game, Network::Ensemble::DIRECT, 0, true);
+            } else if (rotation == "average" || rotation == "avg") {
+                vec = Network::get_scored_moves(
+                    &game, Network::Ensemble::MULTI_AVG, 8, true);
+            } else {
+                vec = Network::get_scored_moves(
+                    &game, Network::Ensemble::DIRECT, std::stoi(rotation), true);
+            }
```

Comment:
Can we get some invalid-argument handling here? Right now, anything that's not a number between 0 and 8 crashes the program, either with an uncaught exception or a segmentation fault.

Reply:
It's always been that way. I thought about trying to handle it but wasn't sure on the best way to do so. Any suggestions?

Reply:
I don't like ```std::stoi``` throwing an exception, as it makes handling both nonintegral as well as out-of-bounds arguments unwieldy. I'd say attempt the conversion through a ```std::stringstream```, as the validity test can then be done with a single ```if``` statement.